### PR TITLE
(#4867) - Cache ddoc views and filters

### PR DESCRIPTION
**Pull request by garethbowen**

### CODE REVIEW

File: src/adapter.js [hunk 1]
```diff
@@ -185,6 +185,71 @@ function attachmentNameError(name) {
...
 
+function cacheUpdateRequired(api, cache, designDocName, callback) {
+  cache.seq = cache.seq || 0;
+  var changesOpts = {
+    doc_ids: [ '_design/' + designDocName ],
+    limit: 1,
+    since: cache.seq
+  };
+  api.changes(changesOpts)
+    .then(function(res) {
```
Comment (daleharvey) [inline src/adapter.js]: Small nit, general style is that as long as they can fit we do

```api.changes(changesOpts).then(function (res) { 
   // rest of code here
});
```

Theres a few other examples of this down below

**Comment by garethbowen** [comment]
@daleharvey I've updated the code style as per your comment.

**Comment by daleharvey** [comment]
Awesome thanks, I was supposed to do more of a review earlier sorry

I think this functionality is mildly worrying, creating additional changes handlers could have unforeseen consequences however it sounds like it creates a significant speedup that cant be done externally.

We should be trusting our tests however, its all passing, I would like performance tests to go along with this ideally, but our performance tests are not in a great state so I will focus on adding some for these soon.

I am :+1: but leave it a little for others to review

**Comment by garethbowen** [comment]
I backed away from continuous changes handlers for that reason, and now it's querying for the latest ddoc when the view/filter is requested. This reduces performance slightly when querying but removes the overhead of listening for changes.

With our (admittedly massive) ddoc on our target phones rendering time for a specific page dropped from ~13 seconds to ~3 seconds with this change.

**Comment by daleharvey** [comment]
Ah I should have read better, for some reason I always see `live: true` automatically when I see `changes()`, that looks like a much safer approach.

**Comment by daleharvey** [comment]
Ok I think this is good to go, thanks so much for the PR @garethbowen - https://github.com/pouchdb/pouchdb/commit/0ddeae6b6c44fa12b83c5348df6947e3ce28c28c

**Comment by nolanlawson** [comment]
This all looks great to me, except the I believe the changes to `test.persisted.js` indicate that we have introduced a bug. If it's not possible to run those promises using `Promise.all()`, then that means map/reduce is not robust against situations where users are querying the same ddoc multiple times in parallel. It indicates a race condition if this code doesn't pass without changing the test.

**Pull request closed without merging**